Document migration steps for Endpoint services to request acls

fsat opened the pull request.

fsat committed:
Document migration steps for Endpoint services to request acls

markusjura commented on src/main/play-doc/general/MigrationGuide.md:
````diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
 ```
 
-### Service ports
+## Service ports
````
What about: "Endpoint declaration"

Sounds more intuitive for me.

edwardcallahan replied:
"Service endpoints" or "Service endpoint declaration"
It's more than just ports

markusjura commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+## Service ports
 
-**The services is now deprecated in lieu of [request acl](RequestAclConfiguration).**
+The services define the protocol, port, and/or path under which your service will be addressed to the outside world on.
```
"The endpoints" instead of "The services" (in case we change the headline).

markusjura replied:
Is the last word "on" necessary?

markusjura commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+The services define the protocol, port, and/or path under which your service will be addressed to the outside world on.
 
-The services define the protocol, port, and/or path under which your service will be addressed to the outside world on. For example, if http and port 80 are to be used to provide your services and then the following expression can be used to resolve `/myservice` on:
+**From 1.2 onwards, the service ports are deprecated in lieu of [[request acl|RequestAclConfiguration]].**
```
[[request ACLs|AclConfiguration]] instead of [[request acl|RequestAclConfiguration]].

markusjura commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
-        protocol  = "http"
-        bind-port = 0
-        services  = ["http:/myservice"]
+The [[request acl|RequestAclConfiguration]] in conjunction with [[dynamic proxy configuration|DynamicProxyConfiguration]] provides additional proxy mapping flexibility on top of what was previously provided by the service ports. This will allow operations to have a complete control on the proxy configuration.
```
[[request acl|AclConfiguration]] instead of [[request acl|RequestAclConfiguration]]

markusjura commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+
+### Prerequisite
+
+Specify the ConductR target version in the `build.sbt`
```
Add a colon at the end.

markusjura commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
-        services  = ["http:/myservice"]
+The [[request acl|RequestAclConfiguration]] in conjunction with [[dynamic proxy configuration|DynamicProxyConfiguration]] provides additional proxy mapping flexibility on top of what was previously provided by the service ports. This will allow operations to have a complete control on the proxy configuration.
+
+### Prerequisite
```
Not sure if we need this `Prerequisite` section. This section only tells, that the user should set the `conductrTargetVersion` to 1.2. Once ConductR 1.2 is released we will change `sbt-conductr` so that the `conductrTargetVersion` defaults to 1.2.

If the user need guidance how to use the `conductrTargetVersion` then this guidance should be part of the `sbt-conductr` README.

Therefore I'd remove the `Prerequisite` section here.

markusjura commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+
+### Migrating HTTP-based services
+
+Some HTTP based service definition can be migrated by supplying its [[request acl|RequestAclConfiguration]] equivalent, while others require a customised HAProxy configuration.
```
[[request ACL|AclConfiguration]] instead of [[request acl|RequestAclConfiguration]]

markusjura commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+
+#### Migrating services definition
+
+The following HTTP based service definition can be migrated by supplying its [[request acl|RequestAclConfiguration]] equivalent.
```
[[request ACL|AclConfiguration]] instead of [[request acl|RequestAclConfiguration]]

markusjura commented on src/main/play-doc/general/MigrationGuide.md:
````diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+
+```
+BundleKeys.endpoints := Map(
+  "endpoint-label" -> Endpoint("http", 0, Set(URI("http:/service-name")))
````
I'd use `my-service` instead of `service-name` to indicate that this can be an arbitrary service name. With this change the above description is also better readable:

"Given the following example where the service URI was "http:/my-service", ConductR will interpret the first path component (`my-service`) as the service name.."

Using `my-service` instead of `service-name` should be applied for all examples in this page.

edwardcallahan replied:
Agreed, this is a confusion causing area and requires careful examples. Also double / in URI string.

markusjura commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+      "endpoint-label" = {
+        bind-protocol = "http"
+        bind-port     = 0
+        services      = ["http:/service-name]
```
Use `my-service` here as well instead of `service-name`.

markusjura commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+##### Services mapped with preserved path
+
+
+Given the following service declaration, the applications or services expect the context root to be preserved when passing through ConductR's proxy, then the service URI can be instead expressed as "http:/service-name?preservePath"
```
What about: "The `preservePath` query parameter on a service declaration indicated that the context root path should be preserved when passing through ConductR's proxy. Given the service URI `http:/my-service?preservePath`, the service is accessible from the outside world on `http://{port-allocated-by-conductr}/my-service`.

This behaviour can now be declared with a request ACL."

edwardcallahan replied:
Paths are preserved by default now?

markusjura commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+The [[Dynamic proxy configuration|DynamicProxyConfiguration]] describes the process of generating a customised HAProxy configuration.
+
+
+##### A note on system, system version, and endpoint label
```
I'd remove this headline. The text of this headline can be part of the previous headline `#### Supplying customised HAProxy template configuration`.

markusjura commented on src/main/play-doc/general/MigrationGuide.md:
````diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+    }
+```
+
+#### Supplying customised HAProxy template configuration
````
`customized` instead of `customised`. In our documentation we are using american english.

markusjura commented:
Generally, use american english in the documentation.

**Examples**
- customized

markusjura commented on src/main/play-doc/general/MigrationGuide.md:
````diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+```
+
+##### Host header match
+
````
Can you describe what a `Host header match` is.

fsat replied:
@markusjura : explanation of host header match provided below

markusjura commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+When declaring bundle settings via `build.sbt`, unless declared explicitly, the `system` and `systemVersion` settings may fallback to a default value as described by [sbt-conductr settings](https://github.com/typesafehub/sbt-conductr#bundle-settings).
+
+
+##### Customised port number
```
Can you add a sentence what a `Customized port number` is.

markusjura commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+
+### Migrating TCP-based services
+
+TCP based service can be migrated by supplying its [[request acl|RequestAclConfiguration]] equivalent.
```
[[request ACL|AclConfiguration]] instead of [[request acl|RequestAclConfiguration]]

markusjura commented:
Should we add an example how to migrate a  `http` service declaration with a customized port:

```
BundleKeys.endpoints := Map(
  "my-endpoint" -> Endpoint("http", 0, Set(URI("http://:5005/my-service")))
)
```

This is probably the most common service declaration used.

edwardcallahan commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+##### Services mapped to root path
+
+Given the following example where the service URI was "http:/service-name", ConductR will interpret the first path component (`service-name`) as the service name for the purposes of service lookup. By default ConductR will then remove the first path component when rewriting the request. This means that your application or service will receive everything under the root of `/`.
+
```
deployed as or in the "root context" is what I 'hear' here.

edwardcallahan commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,333 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+      "endpoint-label" = {
+        bind-protocol = "http"
+        bind-port     = 0
+        services      = ["http:/service-name?preservePath"]
```
http://

fsat commented:
Still under PR review - need to squash commits afterwards

fsat commented on src/main/play-doc/general/MigrationGuide.md:
````diff
@@ -33,18 +33,325 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
 ```
 
-### Service ports
+## Service endpoint declaration
````
@markusjura & @edwardcallahan: header modified

fsat commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,325 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+## Service endpoint declaration
 
-**The services is now deprecated in lieu of [request acl](RequestAclConfiguration).**
+The services endpoint declaration defines the protocol, port, and/or path under which your service will be addressed to the outside world.
```
@markusjura and @edwardcallahan : description updated

fsat commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,325 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
-        services  = ["http:/myservice"]
+The [[request ACL|AclConfiguration]] in conjunction with [[dynamic proxy configuration|DynamicProxyConfiguration]] provides additional proxy mapping flexibility on top of what was previously provided by the service ports. This will allow operations to have a complete control on the proxy configuration.
+
+### Migrating HTTP-based services
```
@markusjura : pre-requisite section removed

fsat commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,325 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+
+##### Services mapped to root path
+
+Given the following example where the service URI was "http://my-service", ConductR will interpret the first path component (`my-service`) as the service name for the purposes of service lookup. By default ConductR will then remove the first path component when rewriting the request. This means that your application or service will be deployed as root context or `/`.
```
@markusjura and @edwardcallahan : `my-service` is now used as the value of the service name.

Also mention service deployed to root context or `/` in this scenario.

fsat commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,325 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+
+##### Services mapped with preserved path
+
+The `preservePath` query parameter on a service declaration indicated that the context root path should be preserved when passing through ConductR's proxy. Given the service URI `http://my-service?preservePath`, the service is accessible from the outside world on HTTP port allocated by ConductR under `/my-sevice` context.
```
@edwardcallahan : paths are not preserved by default - the default behaviour is to have declared service URI rewritten to root context

edwardcallahan replied:
Is this default behavior stated somewhere? That seems useful to know.

fsat commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,325 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+
+The [[Dynamic proxy configuration|DynamicProxyConfiguration]] describes the process of generating a customized HAProxy configuration.
+
+Generating a customized HAProxy configuration will rely on the [[ifAcl|DynamicProxyConfiguration#ifAcl]] directive which accepts `system`, `systemVersion`, and `endpointLabel` as input argument.
```
@markusjura : headline removed

fsat commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,325 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+
+##### Customized port number
+
+A service may declare an endpoint exposed on a custom port number. Given the service URI `http://:6789/my-service`, the service is accessible from the outside world on HTTP port `6789` under `/my-service` context.
```
@markusjura : explanation for customized port number - this is the most common case, i.e. `http://:6789/my-service`

fsat commented:
@markusjura and @edwardcallahan - additional updates: 
- Fixed incorrect link to ACL configuration page
- Fixed incorrect service URI (i.e. ensure it starts with `http://`)
- Replaced "customised" with "customized" (US english)

edwardcallahan commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,325 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+The services endpoint declaration defines the protocol, port, and/or path under which your service will be addressed to the outside world.
 
-The services define the protocol, port, and/or path under which your service will be addressed to the outside world on. For example, if http and port 80 are to be used to provide your services and then the following expression can be used to resolve `/myservice` on:
+**From 1.2 onwards, the service ports are deprecated in lieu of [[request ACL|AclConfiguration]].**
```
'service endpoints are deprecated' vs 'service ports' to be consistent with header

edwardcallahan commented:
Looking good, @fsat.

markusjura commented on src/main/play-doc/general/MigrationGuide.md:
````diff
@@ -33,18 +33,325 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+    }
+```
+
+The following customized template will provide the equivalent proxy mapping:
````
For me it is unclear if the entry in the `build.sbt` / `bundle.conf` needs to be removed or not when creating a customized template?

If I still need to declare an endpoint in the `build.sbt` / `bundle.conf`, shouldn't I then use the new RequestACL format instead of using the "deprecated" services declaration. This should be documented here.

markusjura replied:
Considering a HTTP service declaration with a service name and custom port is the most common use case, shouldn't we offer an API in the `build.sbt` / `bundle.conf` to declare an endpoint without the need of creating a custom template manually?

markusjura commented:
@fsat All changes looking good, thanks. Added two additional comments.

markusjura commented:
Also, we should explain where to put the new custom haproxy template (in the bundle configuration) and provide a documentation link that explains how to create and load a bundle with a configuration. 

This explanation should be part of the headline `Supplying customized HAProxy template configuration`.

fsat commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,425 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+The services endpoint declaration defines the protocol, port, and/or path under which your service will be addressed to the outside world.
 
-The services define the protocol, port, and/or path under which your service will be addressed to the outside world on. For example, if http and port 80 are to be used to provide your services and then the following expression can be used to resolve `/myservice` on:
+**From 1.2 onwards, the service endpoint declaration is deprecated in lieu of [[request ACL|AclConfiguration]].**
```
@edwardcallahan : description updated to match heading

fsat commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,425 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+
+A service may declare an endpoint exposed on a custom port number. Given the service URI `http://:6789/my-service`, the service is accessible from the outside world on HTTP port `6789` under `/my-service` context.
+
+###### Migrating build.sbt
```
@markusjura - I have created a more detailed description on how to migrate the `build.sbt` and `bundle.conf` which was missing earlier - thanks for pointing this out.

fsat commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,425 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+
+A service may declare an endpoint exposed on a particular host address. Given the service URI `http://www.acme.com`, the service is accessible from the outside world on HTTP port `80` when the HTTP host header is `www.acme.com`.
+
+###### Migrating build.sbt
```
@markusjura - I have created a more detailed description on how to migrate the `build.sbt` and `bundle.conf` which was missing earlier - thanks for pointing this out.

fsat commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,425 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+
+The [[Dynamic proxy configuration|DynamicProxyConfiguration]] describes the steps required to generate a customized HAProxy configuration.
+
+Populate the "Custom Endpoint Configuration" section described in the [[populate custom HAProxy configuration template|DynamicProxyConfiguration#Populate-the-custom-HAProxy-configuration-template]] step as such:
```
@markusjura - added link to steps required to supply customized HAProxy configuration

fsat commented on src/main/play-doc/general/MigrationGuide.md:
```diff
@@ -33,18 +33,425 @@ In additional to disabling Elasticsearch via configuration, a service lookup mus
+
+The [[Dynamic proxy configuration|DynamicProxyConfiguration]] describes the steps required to generate a customized HAProxy configuration.
+
+Populate the "Custom Endpoint Configuration" section described in the [[populate custom HAProxy configuration template|DynamicProxyConfiguration#Populate-the-custom-HAProxy-configuration-template]] step as such:
```
@markusjura - added link to steps required to supply customized HAProxy configuration

markusjura commented:
LGTM. Thanks for all the updates.

fsat commented:
@markusjura - it's finally merged! Thanks for the detailed review - enjoy NYC! :-)
